include/circt/Support/PrettyPrinter.h

```diff
@@ -72,6 +72,14 @@ class Token {
   struct EndInfo : public TokenInfo {
     // Nothing
   };
+  // This can be used to associate a callback with the print event on the
+  // tokens stream. Note that the lambda used to create the function object will
+  // be out of scope when it is evoked. So extra care is required to ensure the
+  // values captured by the function object are valid out of scope.
+  struct CallbackInfo : public TokenInfo {
+    using CallbackTy = std::function<void()>;
```

Comment:
std::functions are copy-constructable and fixed size.  If they are capturing something large (like maybe a lambda), they allocate heap space to store them.  I don't think there is a reason to have indirection here.  It doesn't help with lifetimes and it probably hurts performance.  The only reason to would be if the std::function was much larger than a pointer (I think it's 3?, so no) and this was the largest Token type and was uncommon (so making it smaller helped the common case).

I see a union below, perhaps you made this a pointer to get around type safety issues?  If so, maybe use a std::variant?  (or llvm's equivalents, I don't remember the coding guideline on std::variant).

Reply:
It is for type safety issues, I will try other options.